### ISSUE TITLE
Better logic to select which songs belong to the same album.

### ISSUE THREAD

**Issue by p0lygun** [tranxuanthang/lrcget#85]
From what i can see the app uses artist + album ID tag ( i havent looked at the code yet), but this way has many edge cases as many people use [/ (slash) notation](https://community.mp3tag.de/t/how-to-separate-multiple-artist/52232) to specify multiple artists on the same track. (see IMG 1 for reference)

![image](https://github.com/tranxuanthang/lrcget/assets/22869882/83d0aa7c-c61d-4bb4-8371-fdef802424b7)

This leads to the each track being split into an album of its own.

Proposed Solution, 
Various checks can be done to properly club songs together into an album, my flow would be
1. Scan all songs and make a set of values of Album ID3 tag.
2. For each value in set, create an empty album based on Album ID3 tag.
3. Add all songs that have same Album name and Album Artist ID3 tag to this album.
4. If any songs left (globally), Parse Artist name for slash notation, and extract the Artist(s) :
    -  if the first extracted artist name is same as album artist in step 2 and album name is same, add to the same album.
    - else Treat it as a separate album based on Title of the song.

I'd love to contribute to the project, are there any docs on project architecture ?

**Comment by tranxuanthang** [comment]
I totally agree that we should add support for tracks with multiple artists, so they show up correctly in the Albums and Artists tabs. However, be aware of the following:

- The ID3v2.4 tag natively supports multiple artists (separated by a NULL byte). [lofty](https://docs.rs/lofty/latest/lofty/), the library used for processing metadata in LRCGET, should be able to parse that nicely.
- ID3v2.3 doesn't support multiple values at all.

While some songs with ID3v2.3 tags use a slash `/` or comma `,`, we can't simply split the artist field with these characters. You will be surprised by how many artists have a `/` in their nickname, for example: `Au/Ra`, `K/DA`, etc. I think we should provide support only for ID3v2.4, and look more at some other music players to see how they did for ID3v2.3 multiple artists.

Any help is greatly appreciated! There isn't any comprehensive guide about the project architecture yet. But in short, it is made with [Tauri](https://tauri.app/), so:

- There is a frontend part (made with Vue 3) at `/src` directory
- There is a backend part written in Rust at `/src-tauri/src` directory

The frontend, UI part is rendered in a Webview, and it cannot directly interact with filesystem or any special permissions, so that's why the backend is needed. The backend can directly interact with the file system, is a lot faster (so it is used to glob the files in a directory or parse the track's metadata).

For specifically this issue, you can start at the below location. This is where the metadata parsing happening:

https://github.com/tranxuanthang/lrcget/blob/155d53224c2005c3fdc2905252ea9e2efc814ee2/src-tauri/src/fs_track.rs#L65

**Comment by p0lygun** [comment]
ok ty for the heads up, this should be enough for me to start working, i'll ping you  if i need more clarifications

**Comment by ginjaninja1** [comment]
Please consider benefits of supporting Musicbrainz/Picard standard.
https://picard-docs.musicbrainz.org/en/variables/tags_basic.html
Musicbrainz/Picard is a beacon for conformity.
In particular it offers support for disambiguation for multiple artists even when the underlying tag schema does not support multivalue fields.
MBZ's external IDs and API and Database is an opportunity for clarity.

example
![image](https://github.com/tranxuanthang/lrcget/assets/4178240/26c5cd49-1b29-4586-8921-5e76032610cb)

**Comment by ginjaninja1** [comment]
> @ginjaninja1 Thank you, I also tried Musicbrainz Picard lately. I will see how I can use metadata from songs tagged by Picard. I'm also a bit short of time at the moment, so any researches/PRs for this is appreciated!

**Album Appreciation**
I believe correct Album appreciation is more about UI presentation + finding content  rather  than robust lyric lookup/matching.
reason: Lyrics are 'primarily' associated to a track + track length + Artist(s).

Example 'Sam Smith' track on a "Now Thats What I Call Music" compilation album [AlbumArtist=Various Artists] is still the Sam Smith track in terms of fetching lyrics.
Still agree with OP that proper Album presentation is important, just thought its worth highlighting a reason.

**Identifying Unique Albums**
I agree with OP using tags colloquially know as Album Name + Album Artist would be the  way to identify unique albums. (Whilst considering track numbers and Album Disk Number grouping in presentation).
This is the logic i would use.
(For tracks which do not have an AlbumArtist(s) tag but have an Artist(s) tag then AlbumArtist=Artist.)
Tracks with a [derived] AlbumArtist and with Album name - grouped into Albums of the Album Artist.
Tracks with a [derived] AlbumArtist and without Album Name - grouped into "Unknown Album" of the Album Artist
Tracks without a [derived] AlbumArtist and with Album name - grouped in Albums of "Unknown [Album]Artist"
Tracks without a [derived] AlbumArtist and without Album name - grouped in "Unknown Album" of "Unknown [Album]Artist"

Its worth mentioning that some albums have eg 2 Album Artists. When looking at a specific Album it might help to know all the album artists, but in terms of navigation, the album should probably 'appear' under both AlbumArtists
![image](https://github.com/tranxuanthang/lrcget/assets/4178240/6599068c-375d-4b8d-97e6-1c83fb2beb69)

There is always the challenge of supporting users who do not fully tag their tracks.
These users often expect folder structure to define album. (I am not a strong flag waver for such people - if it were up to me i would tell users to 'tag their files')
Driving LRCGET by tag will likely be the gold standard.  But if requiring users to tag seems 'a step too far' then it might be necessary to support a "view by tag/view by folder" 'toggle'. "View by folder" might recommend/require users to adhere to a Root\[Album]Artist\Album[\Disk] file structure. (Its possible that "view by folder users" would would also want all/more metadata to be inferred by the structure/filesystem when not present in tag eg artist and album but perhaps even trackname from filename - ie not just a ui presentation issue but also a metching/fetching issue)

off-topic...
not sure where im going with this....braindump....

**Multiple Artists in a track -Tag Extraction**
MusicBrainz database search can provide list of artists with common separators (/,\,;) in their name for a whitelist.

I suggest the following tag preference for determining the artist lookup for lyric match (and artist extraction for 'track view')
Artist is preferred to AlbumArtist
ArtistS is preferred to Artist. AlbumartistS is preferred to AlbumArtist
I dont know if its pertinent but the first value (Artist) in a multi value field is IMO more significant than subsequent ones.
Generally taggers will put the primary artist first and 'featuring' artists later.

Maybe in time a case could be made for LRCGET database to have a place for MusicBrainz's track and artist ids but i dont see it yet.
It would be rare/exceptional for multiple artists to share the same name whilst performing different songs of the same title (and of the same track length)
LRCGET could go to the nth degree and respect the MUSICBRAINZ_ARTISTID, This would ensure 100% correct matching (for artists with the same string name) but this sounds 'over engineered'.
Support for ArtistS and AlbumArtistS tags sounds like a happy medium.

Note Musicbrainz API will return 'aliases' for an artist and vice versa...this could be useful for fall back matching.

**Multiple Artists in a track -Role in lyric matching.**
Its unclear how additional artists would materially affect lyric identification and matching.
I would guess that a match on any artist (+ track length + track title) should suffice.
When would tracks

- "Don't go breaking my heart" by Elton John + Kiki Dee @ length 4:35 +/-
- "Don't go breaking my heart" by Kiki Dee + Elton John @ length 4:35 +/-
- "Don't go breaking my heart" by Elton John  @ length 4:35 +/-
- "Don't go breaking my heart" by Kiki Dee @ length 4:35 +/-

be different in terms of timed lyrics?

If these tracks really were different (in terms of correct lyric matching) they would have different lengths. I guess it comes down to sort order and preferred match.

Suggested Preferred Order

- All Artists
- First Artist
- Any Artist

Perhaps this topic is as much about how much flexibility is afforded to the relationships to a lyric file in the database i.e.
support for multiple artists and possibly primary artist vs additional artists...

**Comment by tranxuanthang** [comment]
Support for album artist and track number has been added in this commit: https://github.com/tranxuanthang/lrcget/commit/71fa8c0aa84d8a71eeb823a8b06226eee06fb426. This means compilation albums with various artists are now grouped together nicely and displayed in the correct order, too!

**Comment by SomeonetookSomeonetookmeep** [comment]
Right now might actually be a great point to try and get the musicbrainz integration/standard up as they just put out the beta for version 3 of picard, with a new API and some more options. 

Still a pain for true linking, but decent for after. https://blog.metabrainz.org/2026/01/27/picard-3-alpha-release/ (https://github.com/metabrainz/picard/tree/master/docs/PLUGINSV3) post, and the documentation.
